syz-manager/rpc.go

```diff
@@ -366,3 +364,147 @@ func (runner *Runner) newRequest(req *fuzzer.Request) rpctype.ExecutionRequest {
 		NeedHints:    req.NeedHints,
 	}
 }
+
+type requestDistributor struct {
+	fuzzer           *fuzzer.Fuzzer
+	statReusedSafe   *stats.Val
+	statReusedUnsafe *stats.Val
+	// These are the inputs that were sent to VMs, but were not executed.
+	safeInputs   *queue[fuzzer.Request] // from stopped VMs
+	unsafeInputs *queue[fuzzer.Request] // from crashed VMs
+
+	// We don't want to rescue unsafe inputs twice.
+	rescuedUnsafe sync.Map
+}
+
+func newRequestDistributor(fuzzerObj *fuzzer.Fuzzer, statReusedSafe, statReusedUnsafe *stats.Val) *requestDistributor {
+	return &requestDistributor{
+		fuzzer:           fuzzerObj,
+		safeInputs:       newQueue[fuzzer.Request](1000),
+		unsafeInputs:     newQueue[fuzzer.Request](1000),
+		statReusedSafe:   statReusedSafe,
+		statReusedUnsafe: statReusedUnsafe,
+	}
+}
+
+func (rd *requestDistributor) Next(runner *Runner) *fuzzer.Request {
+	if rd.mayUnsafe(runner) {
+		req := rd.unsafeInputs.Fetch()
+		if req != nil {
+			rd.statReusedUnsafe.Add(1)
+			rd.rescuedUnsafe.Store(req, true)
+			return req
+		}
+	}
+	// Next() must be as fast as possible, so let's avoid contention.
+	req := rd.safeInputs.TryFetch()
+	if req != nil {
+		rd.statReusedSafe.Add(1)
+		return req
+	}
+	return rd.fuzzer.NextInput()
+}
+
+func (rd *requestDistributor) mayUnsafe(runner *Runner) bool {
+	if !runner.mu.TryLock() {
+		return false
+	}
+	defer runner.mu.Unlock()
+	uptime := time.Since(runner.started)
+
+	// We don't want to risk crashing freshly booted VMs.
+	// So let's limit the time wasted on VM re-creation to 10% of the total uptime.
+	if (runner.bootTime+uptime).Seconds()*0.1 < runner.bootTime.Seconds() {
```

Comment:
We need to cap bootTime as it may > 1/10-th of an Hour. For proxy VMs we may also account some wait time as booting (e.g. waiting for available machines in a pool).

I also can't immediately comprehend this form. Can it be expressed as `uptime < k*bootTime`?

Reply:
> For proxy VMs we may also account some wait time as booting (e.g. waiting for available machines in a pool).

Do we see this waiting time in `syz-manager` other than as being part of `vmPool.Create()`?

> Can it be expressed as uptime < k*bootTime?

Yes, for 10% that will be `uptime < 9 * bootTime` then.

Reply:
> Do we see this waiting time in syz-manager other than as being part of vmPool.Create()?

No, I think everybody outside of vm/proxy code will just see that as boot time.

I would assume typical boot time for GCE with KASAN kernel is somewhere around 3 mins. If the threshold is 10x, this leaves quite narrow window between 30 mins and 60 mins (max VM run time). For VMs that boot longer we will need something different (e.g. allow unsafe after the same 30 mins). So I wonder perhaps we could just use `time.Since(runner.bootFinished) > X * time.Minutes`.